packages/vhd-lib/src/Vhd/VhdSynthetic.js

```diff
@@ -0,0 +1,86 @@
+import { asyncMap } from '@xen-orchestra/async-map'
+import { VhdAbstract } from './VhdAbstract'
+import { FOOTER_SIZE, HEADER_SIZE, SECTOR_SIZE } from '../_constants'
+import { sectorsRoundUpNoZero, sectorsToBytes } from './_utils'
+
+import assert from 'assert'
+
+export class VhdSynthetic extends VhdAbstract {
+  #vhds = []
+  set header(_) {
+    throw new Error('Header is read only for VhdSynthetic')
+  }
+
+  get header() {
+    // this the VHD we want to synthetize
+    const vhd = this.#vhds[0]
+
+    // this is the root VHD
+    const rootVhd = this.#vhds[this.#vhds.length - 1]
+
+    // data of our synthetic VHD
+    // TODO: set parentLocatorEntry-s in header
+    return {
+      ...vhd.header,
+      tableOffset: FOOTER_SIZE + HEADER_SIZE,
+      parentTimestamp: rootVhd.header.parentTimestamp,
+      parentUnicodeName: rootVhd.header.parentUnicodeName,
+      parentUuid: rootVhd.header.parentUuid,
+    }
+  }
+
+  set footer(_) {
+    throw new Error('Footer is read only for VhdSynthetic')
+  }
+
+  get footer() {
+    // this is the root VHD
+    const rootVhd = this.#vhds[this.#vhds.length - 1]
+    return {
+      ...this.#vhds[0].footer,
+      dataOffset: FOOTER_SIZE,
+      diskType: rootVhd.footer.diskType,
+    }
+  }
+
+  static async open(vhds) {
+    const vhd = new VhdSynthetic(vhds)
+    return {
+      dispose: () => {},
+      value: vhd,
+    }
+  }
+
+  constructor(vhds) {
```

Comment:
Please document the param and be extremely clear of the order of VHDs in it (something I have not done properly when creating `createSyntheticStream`).

Reply:
Also, the idea of synthetic stream is to convert a chain of VHDs (one or more differencing, and a dynamic/differencing parent) into a single VHD.

The result should be similar to a merge operation.

I would expect to see some assertions for this.

Reply:
> I would expect to see some assertions for this.

Shouldn't we assert the types of the VHDs?

Reply:
assertion are done in `readHeaderAndFooter`

Reply:
My bad.